nemo/collections/tts/modules/submodules.py

```
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        style_emb = style_emb.permute(1, 0, 2).contiguous().view(bs, -1)
+
+        return style_emb
+
+
+class SVEmbeddingLinearProjectionNetwork(torch.nn.Module):
+    def __init__(self, speaker_embedding_dim, symbols_embedding_dim):
+        super(SVEmbeddingLinearProjectionNetwork, self).__init__()
+        self.speaker_proj = torch.nn.Linear(speaker_embedding_dim, symbols_embedding_dim)
+        self.act_fn = torch.nn.ReLU()
+
+    def forward(self, sv_embedding):
+        x = self.speaker_proj(sv_embedding)
+        return x.unsqueeze(1)
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+
+    def __init__(self, gst_module=None, sv_projection_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - sv_projection_module: Neural module to get Speaker Verification based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.sv_projection_module = sv_projection_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
+
+    def forward(self, spk_emb, reference_spec=None, reference_spec_lens=None, speaker_embedding=None):
+        x = spk_emb
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            x += self.lookup_emb_projection_module(spk_emb)
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
```

Should we throw an error if gst_module is not None but reference spec is None?

---

Add the check logic in `models/fastpitch.py` https://github.com/NVIDIA/NeMo/blob/tts_adapters_cln/nemo/collections/tts/models/fastpitch.py#L151-L176